amaranth/hdl/_ast.py

```diff
@@ -399,222 +440,595 @@ def __init__(self, *, src_loc_at=0):
         super().__init__()
         self.src_loc = tracer.get_src_loc(1 + src_loc_at)
 
+    @abstractmethod
+    def shape(self):
+        """Shape of :pc:`self`.
+
+        Returns
+        -------
+        :ref:`shape-like object <lang-shapelike>`
```

Comment:
This always returns an actual `Shape`; do we want to leave an option to return a `ShapeCastable`?

Reply:
I added a comment about this. We definitely should decide whether this is useful or not and commit to doing it in `Signal` or whatever or never doing it at all.